abe/abe.go

```diff
@@ -0,0 +1,472 @@
+// the cipher. If the AbeKey is properly generated, this is possible iff
+// the rows of the matrix in the key span the vector [1, 1,..., 1]. If this
+// is not possible, an error is returned.
+func (a Abe) Decrypt(cipher *AbeCipher, key *AbeKey) (*bn256.GT, error) {
+	// get a combination alpha of keys needed to decrypt
+	ones := make(data.Vector, len(key.mat[0]))
+	for i := 0; i < len(ones); i++ {
+		ones[i] = big.NewInt(1)
+	}
+	alpha, err := gaussianElimination(key.mat.Transpose(), ones, a.Params.p)
+	if err != nil {
+		return nil, fmt.Errorf("provided key is not sufficient for decryption")
+	}
+
+	ret := new(bn256.GT).Set(cipher.e0)
+	for i := 0; i < len(alpha); i++ {
+		pair := bn256.Pair(key.d[i], cipher.e[cipher.attribToI[key.rowToAttrib[i]]])
+		pair.ScalarMult(pair, alpha[i])
+		pair.Neg(pair)
+		ret.Add(ret, pair)
+	}
+
+	return ret, nil
+}
+
+// BooleanToMsp takes as an input a boolean expression (without a NOT gate) and
+// outputs a msp structure representing the expression, i.e. a matrix whose rows
+// correspond to attributes used in the expression and with the property that a
+// boolean expression assigning 1 to some attributes is satisfied iff the
+// corresponding rows span a vector [1, 1,..., 1] in Z_p, and a vector whose i-th entry
+// indicates to which attribute the i-th row corresponds.
+func BooleanToMsp(boolExp string, p *big.Int) (*Msp, error) {
+	// by the Lewko-Waters Algorithm we obtain a Msp struct with the property
+	// that is the the boolean expression is satisfied iff the corresponding rows
+	// of the msp matrix span the vector [1, 0,..., 0] in Z_p
+	vec := make(data.Vector, 1)
+	vec[0] = big.NewInt(1)
+	msp, _, err := booleanToMspIterative(boolExp, vec, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	// create an invertible matrix that maps [1, 0,..., 0] to [1,1,...,1]
+	perm := make(data.Matrix, len(msp.mat[0]))
+	for i := 0; i < len(msp.mat[0]); i++ {
+		perm[i] = make(data.Vector, len(msp.mat[0]))
+		for j := 0; j < len(msp.mat[0]); j++ {
+			if i == 0 || j == i {
+				perm[i][j] = big.NewInt(1)
+			} else {
+				perm[i][j] = big.NewInt(0)
+			}
+
+		}
+	}
+	//change the msp matrix so that the boolean expression is satisfied
+	// iff the corresponding rows span the vector [1, 1,..., 1] in Z_p
+	msp.mat, err = msp.mat.Mul(perm)
+	msp.mat = msp.mat.Mod(p)
+	msp.p = p
+	return msp, err
+}
+
+// booleanToMspIterative iteratively builds a msp structure by splitting the expression
+// into two parts separated by an AND or OR gate, generating a msp structure on each of
+// them, and joining both structures together. The structure is such the the boolean expression
+// assigning 1 to some attributes is satisfied iff the corresponding rows span a vector
+// [1, 0,..., 0]. The algorithm is known as Lewko-Waters algorithm, see Appendix G in
+// https://eprint.iacr.org/2010/351.pdf.
+func booleanToMspIterative(boolExp string, vec data.Vector, c int) (*Msp, int, error) {
```

Comment:
Isn't the function both iterative and recursive?

Reply:
True that.